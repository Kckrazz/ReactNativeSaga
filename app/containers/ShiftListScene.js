import React, { Component } from "react";
import {
  Text,
  View,
  StyleSheet,
  BackHandler,
  TouchableOpacity,
  Platform
} from "react-native";
import { Agenda } from "react-native-calendars";
import moment from "moment";
import Ionicons from "react-native-vector-icons/Ionicons";
import { connect } from "react-redux";

import { autobind } from "core-decorators";

import actions from "../actions";
import constants from "../constants";
import selectors from "../selectors";
import {
  Container,
  Header,
  Left,
  Button,
  Body,
  Icon,
  Title
} from "native-base";
import ShiftListItem from "../components/ShiftListItem";
import LocationServicesDialogBox from "react-native-android-location-services-dialog-box";
export class ShiftListScene extends Component {
  static navigationOptions = {
    drawerLabel: "SCHEDULE",
    drawerIcon: ({ tintColor }) =>
      <Ionicons name="md-calendar" style={{ fontSize: 30, color: tintColor }} />
  };
  constructor(props) {
    super(props);
    this.state = {
      current: moment()
    };
  }
  @autobind
  changeDay(day) {
    let d = new Date(day.dateString);
    let date = moment(d);
    this.setState({
      current: date
    });
  }

  componentDidUpdate() {
    // var items = this.props.shifts.toJS();
    // console.log(this.props.shifts.toJS());
  }

  componentWillUnmount() {
    navigator.geolocation.clearWatch(this.watchId);
  }

  async checkIsLocation(): Boolean {
    let check = await LocationServicesDialogBox.checkLocationServicesIsEnabled({
      message:
        "<h2>Use Location?</h2> \
                            This app wants to change your device settings:<br/><br/>\
                            Use GPS for location<br/><br/>",
      ok: "YES",
      cancel: "NO"
    })
      .then(success => {
        const action = actions.location.locationRequest();
        this.props.dispatch(action);
      })
      .catch(error => {
        console.log(error.message);
      });

    return Object.is(check, "enabled");
  }
  componentDidMount() {
    if (Platform.OS === "android") {
      this.checkIsLocation();
    }
    const action = actions.location.locationRequest();
    this.props.dispatch(action);
  }

  render() {
    let title = this.state.current.format("YYYY-MM-DD");
    return (
      <Container>
        <Header style={{ backgroundColor: "#414141" }}>
          <Left>
            <Button
              transparent
              onPress={() => this.props.navigation.navigate("DrawerOpen")}
            >
              <Icon name="menu" />
            </Button>
          </Left>
          <Body>
            <Title>
              {this.state.current.format("MMM YYYY").toUpperCase()}
            </Title>
          </Body>
        </Header>
        <Agenda
          ref={ref => {
            this.agenda = ref;
          }}
          items={this.props.shifts.toJS()}
          loadItemsForMonth={this.loadItems}
          selected={title}
          onDayPress={this.changeDay}
          onDayChange={this.changeDay}
          theme={{}}
          // agenda container style
          style={{}}
          renderItem={this.renderItem}
          renderEmptyDate={this.renderEmptyDate}
          rowHasChanged={this.rowHasChanged}
          //renderDay={(day, item) => (<Text>{day ? day.day: 'item'}</Text>)}
        />
      </Container>
    );
  }

  @autobind
  loadItems(day) {
    const selected_date = moment(new Date(day.dateString));
    const action = actions.shift.loadShiftsRequest({ date: selected_date });
    this.props.dispatch(action);
  }

  @autobind
  _navigate(item) {
    const shift = this.props.shifts;
    const obj = shift.filter(s => {
      if (!s.isEmpty()) {
        let sh = s.toJS();
        let song = _.some(sh, { id: item.id });
        return song ? true : false;
      } else {
        return false;
      }
    });
    this.props.navigation.navigate("Shift", {
      shift: item,
      date: obj.keySeq().first()
    });
  }

  @autobind
  renderItem(item) {
    const _onClick = () => {
      this._navigate(item);
    };
    return (
      <TouchableOpacity onPress={_onClick}>
        <ShiftListItem data={item} />
      </TouchableOpacity>
    );
  }

  @autobind
  renderEmptyDate() {
    return <View style={styles.emptyDate}><Text>No Shifts!</Text></View>;
  }

  @autobind
  rowHasChanged(r1, r2) {
    return r1.title !== r2.title;
  }

  timeToString(time) {
    const date = new Date(time);
    return date.toISOString().split("T")[0];
  }
}

const styles = StyleSheet.create({
  item: {
    backgroundColor: "white",
    flex: 1,
    borderRadius: 5,
    padding: 10,
    marginRight: 10,
    marginTop: 5
  },
  emptyDate: {
    height: 15,
    flex: 1,
    paddingTop: 30
  }
});

const mapStateToProps = state => ({
  busy: selectors.user.authPending(state),
  error: selectors.user.authError(state),
  authenticated: selectors.user.authenticated(state),
  shifts: state.shifts
});

export default connect(mapStateToProps)(ShiftListScene);
