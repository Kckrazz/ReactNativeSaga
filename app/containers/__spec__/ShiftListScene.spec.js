import React from "react";
import "react-native";
import { ShiftListScene } from "../ShiftListScene";
import { Map } from "immutable";
import sinon from "sinon";
import renderer from "react-test-renderer";
describe("ShiftListScene", () => {
  test("Should Render Snapshot Correctly", () => {
    const props = {
      busy: false,
      authenticated: true,
      error: null,
      shifts: new Map({}),
      dispatch: sinon.spy()
    };
    const tree = renderer.create(<ShiftListScene {...props} />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
