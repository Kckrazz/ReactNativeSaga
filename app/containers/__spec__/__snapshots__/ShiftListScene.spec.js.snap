// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ShiftListScene Should Render Snapshot Correctly 1`] = `
<View
  style={
    Object {
      "flex": 1,
      "height": 1334,
    }
  }
  virtual={undefined}
>
  <View>
    <View
      style={
        Object {
          "backgroundColor": "#414141",
          "borderBottomColor": "#a7a6ab",
          "borderBottomWidth": 0.5,
          "elevation": 3,
          "flexDirection": "row",
          "height": 64,
          "justifyContent": "center",
          "left": 0,
          "paddingLeft": 10,
          "paddingRight": 10,
          "paddingTop": 15,
          "right": 0,
          "shadowColor": undefined,
          "shadowOffset": undefined,
          "shadowOpacity": undefined,
          "shadowRadius": undefined,
          "top": 0,
        }
      }
      virtual={undefined}
    >
      <View
        style={
          Object {
            "alignItems": "flex-start",
            "alignSelf": "center",
            "flex": 1,
          }
        }
        virtual={undefined}
      >
        <View
          accessibilityComponentType={undefined}
          accessibilityLabel={undefined}
          accessibilityTraits={undefined}
          accessible={true}
          hitSlop={undefined}
          isTVSelectable={true}
          onLayout={undefined}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "alignItems": "center",
              "alignSelf": null,
              "backgroundColor": "transparent",
              "borderBottomWidth": null,
              "borderColor": null,
              "borderLeftWidth": null,
              "borderRadius": 5,
              "borderRightWidth": null,
              "borderTopWidth": null,
              "elevation": 0,
              "flexDirection": "row",
              "height": 45,
              "justifyContent": "space-between",
              "marginLeft": -3,
              "opacity": 1,
              "paddingBottom": 6,
              "paddingLeft": 6,
              "paddingRight": 6,
              "paddingTop": 6,
              "shadowColor": null,
              "shadowOffset": null,
              "shadowOpacity": null,
              "shadowRadius": null,
            }
          }
          testID={undefined}
          tvParallaxProperties={undefined}
        >
          <Text
            accessible={true}
            allowFontScaling={false}
            ellipsizeMode="tail"
            style={
              Array [
                Object {
                  "color": undefined,
                  "fontSize": 12,
                },
                Object {
                  "color": "#007aff",
                  "fontSize": 33,
                  "marginLeft": 2,
                  "marginRight": 2,
                  "marginTop": 2,
                  "paddingTop": 2,
                },
                Object {
                  "fontFamily": "Ionicons",
                  "fontStyle": "normal",
                  "fontWeight": "normal",
                },
              ]
            }
            virtual={undefined}
          >
            ï‡‚
          </Text>
        </View>
      </View>
      <View
        style={
          Object {
            "alignItems": "center",
            "alignSelf": "center",
            "flex": 1,
          }
        }
        virtual={undefined}
      >
        <Text
          accessible={true}
          allowFontScaling={true}
          ellipsizeMode="tail"
          numberOfLines={1}
          style={
            Object {
              "color": "#000",
              "fontFamily": "System",
              "fontSize": 17,
              "fontWeight": "600",
              "textAlign": "center",
            }
          }
          virtual={undefined}
        >
          JUN 2017
        </Text>
      </View>
    </View>
  </View>
  <View
    onLayout={[Function]}
    style={
      Array [
        Object {},
        Object {
          "flex": 1,
        },
      ]
    }
  >
    <View
      style={
        Object {
          "backgroundColor": "#f4f4f4",
          "flex": 1,
          "marginTop": 104,
        }
      }
    >
      <ActivityIndicator
        animating={true}
        color="#999999"
        hidesWhenStopped={true}
        size="small"
        style={
          Object {
            "marginTop": 80,
          }
        }
      />
    </View>
    <View
      style={
        Object {
          "borderBottomWidth": 1,
          "borderColor": "#e8e9ec",
          "height": 104,
          "left": 0,
          "position": "absolute",
          "right": 0,
          "top": 0,
        }
      }
    >
      <ListView
        dataSource={
          ListViewDataSource {
            "items": 101,
          }
        }
        initialListSize={2501}
        onChangeVisibleRows={[Function]}
        onLayout={[Function]}
        onScroll={[Function]}
        pageSize={1}
        removeClippedSubviews={true}
        renderRow={[Function]}
        scrollEnabled={false}
        scrollRenderAheadDistance={360}
        showsVerticalScrollIndicator={false}
        style={
          Array [
            Object {
              "backgroundColor": "#ffffff",
            },
            undefined,
          ]
        }
      />
      <View
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "#ffffff",
            "bottom": 0,
            "flex": 1,
            "height": 24,
            "left": 0,
            "position": "absolute",
            "right": 0,
          }
        }
      >
        <View
          accessibilityComponentType={undefined}
          accessibilityLabel={undefined}
          accessibilityTraits={undefined}
          accessible={true}
          hitSlop={undefined}
          isTVSelectable={true}
          onLayout={undefined}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "opacity": 1,
            }
          }
          testID={undefined}
          tvParallaxProperties={undefined}
        >
          <View
            style={
              Object {
                "backgroundColor": "#f2F4f5",
                "borderRadius": 3,
                "height": 7,
                "marginTop": 10,
                "width": 38,
              }
            }
          />
        </View>
      </View>
    </View>
    <View
      style={
        Object {
          "backgroundColor": "#ffffff",
          "flexDirection": "row",
          "justifyContent": "space-around",
          "left": 0,
          "marginLeft": 15,
          "marginRight": 15,
          "opacity": 1,
          "paddingBottom": 7,
          "paddingTop": 15,
          "position": "absolute",
          "right": 0,
          "top": 0,
        }
      }
    >
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        style={
          Object {
            "color": "#b6c1cd",
            "fontSize": 13,
            "textAlign": "center",
            "width": 32,
          }
        }
      >
        Sun
      </Text>
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        style={
          Object {
            "color": "#b6c1cd",
            "fontSize": 13,
            "textAlign": "center",
            "width": 32,
          }
        }
      >
        Mon
      </Text>
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        style={
          Object {
            "color": "#b6c1cd",
            "fontSize": 13,
            "textAlign": "center",
            "width": 32,
          }
        }
      >
        Tue
      </Text>
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        style={
          Object {
            "color": "#b6c1cd",
            "fontSize": 13,
            "textAlign": "center",
            "width": 32,
          }
        }
      >
        Wed
      </Text>
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        style={
          Object {
            "color": "#b6c1cd",
            "fontSize": 13,
            "textAlign": "center",
            "width": 32,
          }
        }
      >
        Thu
      </Text>
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        style={
          Object {
            "color": "#b6c1cd",
            "fontSize": 13,
            "textAlign": "center",
            "width": 32,
          }
        }
      >
        Fri
      </Text>
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        style={
          Object {
            "color": "#b6c1cd",
            "fontSize": 13,
            "textAlign": "center",
            "width": 32,
          }
        }
      >
        Sat
      </Text>
    </View>
  </View>
  <Modal
    animationType="fade"
    hardwareAccelerated={false}
    onRequestClose={[Function]}
    supportedOrientations={null}
    transparent={true}
    visible={false}
  >
    <View
      style={
        Object {
          "flex": 1,
          "justifyContent": "flex-start",
          "margin": 20,
        }
      }
    >
      <View
        danger={false}
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "rgba(0,0,0,0.8)",
            "borderRadius": 5,
            "flexDirection": "row",
            "justifyContent": "space-between",
            "minHeight": 50,
            "paddingBottom": 10,
            "paddingLeft": 10,
            "paddingRight": 10,
            "paddingTop": 10,
          }
        }
        success={false}
        virtual={undefined}
        warning={false}
      >
        <Text
          accessible={true}
          allowFontScaling={true}
          ellipsizeMode="tail"
          style={
            Object {
              "color": "#fff",
              "flex": 1,
              "fontFamily": "System",
              "fontSize": 16,
            }
          }
          virtual={undefined}
        />
      </View>
    </View>
  </Modal>
  <Modal
    animationType="fade"
    hardwareAccelerated={false}
    onRequestClose={[Function]}
    transparent={true}
    visible={false}
  >
    <View
      accessibilityComponentType={undefined}
      accessibilityLabel={undefined}
      accessibilityTraits={undefined}
      accessible={true}
      hitSlop={undefined}
      isTVSelectable={true}
      onLayout={undefined}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "backgroundColor": "rgba(0,0,0,0.4)",
          "flex": 1,
          "justifyContent": "flex-end",
          "opacity": 1,
        }
      }
      testID={undefined}
      tvParallaxProperties={undefined}
    >
      <View
        accessibilityComponentType={undefined}
        accessibilityLabel={undefined}
        accessibilityTraits={undefined}
        accessible={true}
        hitSlop={undefined}
        isTVSelectable={true}
        onLayout={undefined}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "backgroundColor": "#fff",
            "elevation": 4,
            "height": NaN,
            "opacity": 1,
            "padding": 15,
          }
        }
        testID={undefined}
        tvParallaxProperties={undefined}
      >
        <Text
          accessible={true}
          allowFontScaling={true}
          ellipsizeMode="tail"
          style={
            Object {
              "color": "#757575",
              "fontFamily": "System",
              "fontSize": 16,
            }
          }
          virtual={undefined}
        />
        <ListView
          dataArray={Array []}
          dataSource={
            ListViewDataSource {
              "items": 0,
            }
          }
          enableEmptySections={true}
          renderRow={[Function]}
          style={
            Object {
              "marginHorizontal": -15,
              "marginTop": 15,
            }
          }
          virtual={undefined}
        />
      </View>
    </View>
  </Modal>
</View>
`;
